dev-tools/omdb/src/bin/omdb/mgs.rs

```diff
@@ -0,0 +1,499 @@
+    Inventory(InventoryArgs),
+}
+
+#[derive(Debug, Args)]
+struct InventoryArgs {}
+
+impl MgsArgs {
+    pub(crate) async fn run_cmd(
+        &self,
+        omdb: &Omdb,
+        log: &slog::Logger,
+    ) -> Result<(), anyhow::Error> {
+        let mgs_url = match &self.mgs_url {
+            Some(cli_or_env_url) => cli_or_env_url.clone(),
+            None => {
+                eprintln!(
+                    "note: MGS URL not specified.  Will pick one from DNS."
+                );
+                let addrs = omdb
+                    .dns_lookup_all(
+                        log.clone(),
+                        internal_dns::ServiceName::ManagementGatewayService,
+                    )
+                    .await?;
+                let addr = addrs.into_iter().next().expect(
+                    "expected at least one MGS address from \
+                    successful DNS lookup",
+                );
+                format!("http://{}", addr)
+            }
+        };
+        eprintln!("note: using MGS URL {}", &mgs_url);
+        let mgs_client = gateway_client::Client::new(&mgs_url, log.clone());
+
+        match &self.command {
+            MgsCommands::Inventory(inventory_args) => {
+                cmd_mgs_inventory(&mgs_client, inventory_args).await
+            }
+        }
+    }
+}
+
+/// Runs `omdb mgs inventory`
+///
+/// Shows devices and components that are visible to an MGS instance.
+async fn cmd_mgs_inventory(
+    mgs_client: &gateway_client::Client,
+    _args: &InventoryArgs,
+) -> Result<(), anyhow::Error> {
+    // Report all the SP identifiers that MGS is configured to talk to.
+    println!("ALL CONFIGURED SPs\n");
+    let mut sp_ids = mgs_client
+        .sp_all_ids()
+        .await
+        .context("listing SP identifiers")?
+        .into_inner();
+    sp_ids.sort_by(sp_id_cmp);
+    show_sp_ids(&sp_ids)?;
+    println!("");
+
+    // Report which SPs are visible via Ignition.
+    println!("SPs FOUND THROUGH IGNITION\n");
+    let mut sp_list_ignition = mgs_client
+        .ignition_list()
+        .await
+        .context("listing ignition")?
+        .into_inner();
+    sp_list_ignition.sort_by(|a, b| sp_id_cmp(&a.id, &b.id));
+    show_sps_from_ignition(&sp_list_ignition)?;
+    println!("");
+
+    // Print basic state about each SP that's visible to ignition.
+    println!("SERVICE PROCESSOR STATES\n");
+    let mgs_client = std::sync::Arc::new(mgs_client);
+    let c = &mgs_client;
+    let mut sp_infos =
+        futures::stream::iter(sp_list_ignition.iter().filter_map(|ignition| {
+            if matches!(ignition.details, SpIgnition::Yes { .. }) {
+                Some(ignition.id)
+            } else {
+                None
+            }
+        }))
+        .then(|sp_id| async move {
+            let msg = format!("fetching info about SP {:?}", sp_id);
+            c.sp_get(sp_id.type_, sp_id.slot)
+                .await
+                .context(msg)
```

Comment:
Tiny nit - as written, we always allocate `msg` but then immediately free it if the request succeeds. Could we use

```suggestion
                .with_context(|| format!("..."))
```

instead, or do we run into problems copying/cloning `sp_id`? (That seems like it should be `Copy`, I think?)

Reply:
Thanks. (This was a remnant of an earlier version where the same message was used in multiple places.)